### PR TITLE
add khmer segmenter

### LINKED ISSUE #200: add support for khmer language
adding support for khmer language, started from this [discussion](https://github.com/meilisearch/product/discussions/374)

### PR DESCRIPTION
**Pull request by xshadowlegendx**
# Pull Request

## Related issue
Fixes #200 

## What does this PR do?
- add segmenter for `khmer` language

## PR checklist
Please check if your PR fulfills the following requirements:
- [x] Does this PR fix an existing issue, or have you listed the changes applied in the PR description (and why they are needed)?
- [x] Have you read the contributing guidelines?
- [x] Have you made sure that the title is accurate and descriptive of the changes?

Thank you so much for contributing to Meilisearch!

### CODE REVIEW

**Comment by ManyTheFish** [comment]
Hello @xshadowlegendx,

don't hesitate to pull the main branch of Meilisearch and update the [milli/Cargo.toml](https://github.com/meilisearch/meilisearch/blob/main/milli/Cargo.toml#L19) file linking your branch instead of a fixed version of Charabia.

This way you will be able to run Meilisearch with your changes!

**Comment by xshadowlegendx** [comment]
@ManyTheFish ok thanks will try it

**Comment by xshadowlegendx** [comment]
hello @ManyTheFish, sorry for the late reply. Currently I am still testing the khmer segmenter with meilisearch, so I am able to compile and run it with local `charabia` as dependency. Is there anything like text analysis where I can test if the segmenter works? so the way I am testing now is inserting some khmer data and search for word for it but it seems to search it part by part instead of word? I also tried to run a meilisearch docker container to compare and it seems to work the same as well, not sure if the segmenter is working or not yet

**Comment by ManyTheFish** [comment]
@xshadowlegendx, a test you could try is to push a document containing `សួស្តីពិភពលោក` and try searching for `ពិភពលោក`. As Meilisearch only supports prefix search then you should not be able to find the document containing `សួស្តីពិភពលោក` if your segmenter is not used. Otherwise, Meilisearch managed to segment `សួស្តីពិភពលោក` and so we can conclude with the fact that your segmenter is used.
If you want to activate some normalizer or add a specific one for the Khmer Language, don't hesitate to [check the normalizer list](https://github.com/xshadowlegendx/charabia/tree/add-khmer-segmenter/charabia/src/normalizer), probably removing diacritics is relevant for the Khmer Language, In this case, I suggest adding it to the [normalizer's language list](https://github.com/xshadowlegendx/charabia/blob/91049d6f254bac63ae6121064a6c1f9469dfc877/charabia/src/normalizer/nonspacing_mark.rs#L31).

I hope it helps!

**Comment by xshadowlegendx** [comment]
ok thanks @ManyTheFish  let me try to test it again

**Comment by xshadowlegendx** [comment]
hello @ManyTheFish , so I tried pushing document that contains `សួស្តីពិភពលោក` and searching for `ពិភពលោក` it is able to find those document that contains it, this is with meilisearch from dockerhub without the khmer segmenter, below are steps I took to test it
```
docker run -it --rm -p 7800:7700 getmeili/meilisearch:v1.2

# books.json
[
  {
    "id": "ec660236154a4bed88b554cff8fb93a4",
    "title": "ភាសាខ្មែរជាភាសល្អជាងគេនៅពិភពលោក"
  },
  {
    "id": "2dea8651215a4695aaf8721f4b08fa40",
    "title": "ប្រទេសចិនប្រេីភាសាចិន"
  },
  {
    "id": "2dea8651215a4695aaf8721f4b08fa41",
    "title": "សួស្តីពិភពលោក"
  }
]

curl -X POST 'http://localhost:7800/indexes/books/documents?primaryKey=id' -H 'Content-Type: application/json' --data-binary @books.js
```

this screenshot seems like without khmer segmenter it also able to pick up the word
Screenshot 2023-06-24 at 11 18 19

but if its not a word it also pick it up
Screenshot 2023-06-24 at 11 18 44

so I try to test the khmer segmenter by cloning meilisearch to my local machine, point charabia deps to my local machine one and `cargo run` to run meilisearch and test just like above. Before the testing, I put a `panic!("khmer segmenter says panic");` within the `segment_str` function to ensure it gets called and it does, then I remove the panic statement and proceed to test like above and got the same result.

below are changes to local meilisearch to test with the khmer segmenter
```
❯ cat milli/Cargo.toml | grep chara
charabia = { path = "../../charabia/charabia", default-features = false }
default = [ "charabia/default" ]
chinese = ["charabia/chinese"]
hebrew = ["charabia/hebrew"]
japanese = ["charabia/japanese"]
japanese-transliteration = ["charabia/japanese-transliteration"]
korean = ["charabia/korean"]
thai = ["charabia/thai"]
khmer = ["charabia/khmer"]
```

File: charabia/icu4x-khmer-keys
Comment (ManyTheFish) [inline charabia/icu4x-khmer-keys:1]: Could you move this file in `charabia/dictionaries/bin/, please?

Comment (xshadowlegendx) [reply charabia/icu4x-khmer-keys:1]: ok will do

**Comment by xshadowlegendx** [comment]
> hello @xshadowlegendx, I asked for small changes in your PR,
> 
>     1. Do these changes enhance your search results?
> 
>     2. Do you see some search results that are missing?
> 
> 

> See you!

after those changes on my local, now it show some differences, the left one is running with docker and the right one is running with local development, but searches does not seem to be correct? so I tried searching the word `លោក` it shows empty result which should not and typing only a character also matches is this expected behavior? and I try the word `ល្អ`, it matches one document which is correct but also matches another one which does not have the word but has the `ល` character in it, which is `លោក` that has the character in the middle

Screenshot 2023-07-18 at 23 39 13

**Comment by ManyTheFish** [comment]
Hello @xshadowlegendx, Meilisearch is a prefix search engine, so it's expected that if you tip the start of a word Meilisearch find all the documents containing a word starting with the query.
Is it the case?

**Comment by xshadowlegendx** [comment]
hello @ManyTheFish , yes that is the case, so it is expected behavior then. I will be testing some document that are more close to real world data to see how the segmenter perform next then

**Comment by ManyTheFish** [comment]
Hello @xshadowlegendx,

Any news on this PR, Do you feel enough comfortable with your tests to merge it?

Thanks!

**Comment by xshadowlegendx** [comment]
hello @ManyTheFish, sorry for the late replies, so I tested with about 253 movies that is production data from my workplace weeks ago. the search is very fast, can search movies from titles, actor names and genres, and typo tolerance also works by updating the default since khmer word are smaller size, so I think it is ready to be merged.

but the current `icu4x` dependency that I am using currently have a bug with using dictionary model but it was fixed very soon after I submit the [issue](https://github.com/unicode-org/icu4x/issues/3545) but It hasnt been release yet

**Comment by curquiza** [comment]
Hello @xshadowlegendx 
Can you fix the git conflicts you have on this PR before we review? 😊

**Comment by curquiza** [comment]
@xshadowlegendx 
Looks like clippy and rust fmt CI fail. Can you fix these before we review?

**Comment by xshadowlegendx** [comment]
hello @curquiza, the `clippy` wants me to fix the `invisible_characters` such as below

```
❯ cargo clippy --all-targets -- --deny warnings
    Checking charabia v0.8.4 (/Users/xanonx/projects/meilisearch/charabia/charabia)
error: invisible character detected
  --> charabia/benches/bench.rs:28:43
   |
28 | ...), "រឿង​ពីរ​ដែល​មនុស្ស​ហាម​ចិត្ត​ខ្លួន​ឯង​មិន​បាន​គឺ​ សើច​ និង​ ស្រឡាញ់​​។ តែ​សម្រាប់​ខ្ញុំ​ ប្រាក់ ចន្ទធីតា​ ​រឿង​មួយ​ទៀត​ដែល​ខ្ញុំ​ហាម​ចិត្ត​ខ្លួន​ឯង​មិន​បាន​នោះ គឺ​ញ៉ាំ​ គេ​គ្រប់​គ្នា​ពេល​ខូច​ចិត...បាយ​ទឹក​មិន​បាន​ទេ តែ​ខ្ញុំ​ពេល​ខូច​ចិត្ត​ដឹង​តែ​ឃ្លាន​ ញ៉ាំ​ច្រើន​ឬ​តិច​ក៏​អាស្រ័យ​​លើ​ថា​ទំហំ​នៃការ​ខូច​​ចិត្ត​ខ្លាំង​ឬ​ខ្សោយ​​។"),
   |       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider replacing the string with: `"រឿង\u{200B}ពីរ\u{200B}ដែល\u{200B}មនុស្ស\u{200B}ហាម\u{200B}ចិត្ត\u{200B}ខ្លួន\u{200B}ឯង\u{200B}មិន\u{200B}បាន\u{200B}គឺ\u{200B} សើច\u{200B} និង\u{200B} ស្រឡាញ់\u{200B}\u{200B}។ តែ\u{200B}សម្រាប់\u{200B}ខ្ញុំ\u{200B} ប្រាក់ ចន្ទធីតា\u{200B} \u{200B}រឿង\u{200B}មួយ\u{200B}ទៀត\u{200B}ដែល\u{200B}ខ្ញុំ\u{200B}ហាម\u{200B}ចិត្ត\u{200B}ខ្លួន\u{200B}ឯង\u{200B}មិន\u{200B}បាន\u{200B}នោះ គឺ\u{200B}ញ៉ាំ\u{200B} គេ\u{200B}គ្រប់\u{200B}គ្នា\u{200B}ពេល\u{200B}ខូច\u{200B}ចិត្ត\u{200B}បាយ\u{200B}ទឹក\u{200B}មិន\u{200B}បាន\u{200B}ទេ តែ\u{200B}ខ្ញុំ\u{200B}ពេល\u{200B}ខូច\u{200B}ចិត្ត\u{200B}ដឹង\u{200B}តែ\u{200B}ឃ្លាន\u{200B} ញ៉ាំ\u{200B}ច្រើន\u{200B}ឬ\u{200B}តិច\u{200B}ក៏\u{200B}អាស្រ័យ\u{200B}\u{200B}លើ\u{200B}ថា\u{200B}ទំហំ\u{200B}នៃការ\u{200B}ខូច\u{200B}\u{200B}ចិត្ត\u{200B}ខ្លាំង\u{200B}ឬ\u{200B}ខ្សោយ\u{200B}\u{200B}។"`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#invisible_characters
   = note: `#[deny(clippy::invisible_characters)]` on by default

error: could not compile `charabia` (bench "bench") due to previous error
```
should I apply the suggested fix?

**Comment by ManyTheFish** [comment]
hello @xshadowlegendx,
if you can't remove the invisible character from the code, then you can add the flag to ignore the error, honestly, I'd prefer it if the invisible character is removed. 😄 

Thanks!

**Comment by ManyTheFish** [comment]
bors try

**Review by ManyTheFish — APPROVED** [review]
Looks good to me,
thank you @xshadowlegendx for your work!

bors merge

**Pull request merged**